_posts/2024-06-28-overriding-configuration-from-test-code.adoc

```diff
@@ -0,0 +1,626 @@
+
+== Approach #5: system properties
+
+I would definitely NOT recommend this approach, but it does exist and it kinda works, so I'll mention it anyway.
+System properties can be used to override the configuration from the test code.
+This approach suffers however from major drawbacks:
+
+* It doesn't work in native mode.
+* It doesn't work with https://quarkus.io/guides/config-mappings[config mappings^].
+* It only works once when the configuration is defined in an `@ApplicationScoped` or `@Singleton` bean, before that bean has been initialized.
+After the bean initialization, any changes made to system properties will have no effect on the configuration.
+
+[source, java, title=The code to be tested]
+----
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+@Path("/features")
+public class FeaturesResource {
+
+    @ConfigProperty(name = "awesome-feature-enabled", defaultValue = "false")
+    boolean awesomeFeatureEnabled;
+
+    @GET
+    @Path("/awesome")
+    public boolean isAwesomeFeatureEnabled() {
+        return awesomeFeatureEnabled;
+    }
+}
+----
+
+System properties can be set from the command line with Maven or Gradle:
+
+[source, shell, title=Maven command]
+----
+./mvnw verify -Dawesome-feature-enabled=true
+----
+
+They can also be set from the test code:
+
+[source, java, title=The test code]
+----
+import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.RestAssured;
+import org.hamcrest.CoreMatchers;
+import org.junit.jupiter.api.MethodOrderer;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+
+@QuarkusTest
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class) // <1>
+class FeaturesResourceTest {
+
+    @Test
+    @Order(1) // <2>
+    void firstTest() {
+        System.setProperty("awesome-feature-enabled", "true");
+        RestAssured.given()
+            .when().get("/features/awesome")
+            .then().body(CoreMatchers.is("true"));
+    }
+
+    @Test
+    @Order(2) // <3>
+    void lastTest() {
+        System.setProperty("awesome-feature-enabled", "false");
+        RestAssured.given()
+            .when().get("/features/awesome")
+            .then().body(CoreMatchers.is("true")); // <4>
+    }
+}
+----
+<1> In this code snippet, tests are run in a fixed order to demonstrate a limitation of system properties.
+<2> This test always runs first.
+<3> This test always runs last.
+<4> This test depends on a CDI bean with a default `@Singleton` scope which was already initialized by the previous test.
+As a consequence, the outcome of this test cannot be changed from the system property.
+
+== Off-topic: the `application.properties` file
+
+You may wonder why I didn't mention the `application.properties` file sooner.
+It can indeed be used to change the configuration of a Quarkus app while testing it.
```

Comment:
You can create profile-aware files: https://quarkus.io/guides/config-reference#profile-aware-files. They can sit in `src/test/resournce` to not polute the main sources, and then you just override `io.quarkus.test.junit.QuarkusTestProfile#getConfigProfile` to load specific files per test.

Reply:
Interesting. I'll add that to the post. Thanks!

Reply:
@radcortez Is your suggestion supposed to work in native mode? I tried with no luck so far... (JVM mode works fine)

Reply:
This might be related:
```
2024-06-29 02:42:32,151 WARN  [io.qua.run.con.ConfigRecorder] (main) The profile 'prod' used to build the native image is different from the runtime profile 'blog'. This may lead to unexpected results.
```

Reply:
I made it work in native mode by moving the properties files from `src/test/resources` to `src/main/resources`.

I also had to add this in `src/main/resources/application.properties`:
```
quarkus.native.resources.includes=application*.properties
```